Add word2vec.PathLineSentences for reading a directory as a corpus (#1364)

michaelwsherman opened the pull request:
word2vec.LineSentencePath(path) will read all the files in a directory in the same fashion as word2vec.LineSentence reads a file. This provides an easy way to use a corpus of multiple files when training a word2vec model (or any model compatible with word2vec.LineSentence). 

Minimal exeception handling, but some logging.

gojomo commented:
Thanks for the contrib!

The failing automatic check looks like some style issues - you can click the 'Details' link or red 'X' to view the failure logs for more hints. 

More substantive comments inline.

gojomo commented on gensim/models/word2vec.py:
```diff
@@ -1521,6 +1521,54 @@ def __iter__(self):
                         i += self.max_sentence_length
 
 
+class LineSentencePath(object):
```
Personally I'd consider the name `PathLineSentences` more descriptive, but other may have an even better name.

michaelwsherman replied:
Thank you for taking the time to comment. I'll get to these next week.

michaelwsherman replied:
Change made, will be reflected in next pull request

gojomo commented on gensim/models/word2vec.py:
```diff
@@ -1521,6 +1521,54 @@ def __iter__(self):
+        self.limit = limit
+
+        try:
+            self.source = os.path.join(source, '') # ensures os-specific slash is at end of path
```
Thoughts on coverage of all related needs:

* perhaps this should accept a path to a single file, too, and still work in that case?
* by deferring the actual resolution of initialization parameters to the beginning of `__iter__()`, the object might be more robust for cases where files are arriving in the target directory between instantiation & 1st iteration. OTOH, that would also mean repeated iterations – as in the common Word2Vec/Doc2Vec multi-pass training, could find different files each time. No strong opinion yet on which approach is better – just pointing out the choice.

michaelwsherman replied:
Change made to accept a single file and still work, including an additional test case.

I think it is better to resolve the initialization parameters in the `__init__()`. While there could be some use in not requiring the files to all be present when the object is initialized, I think that possibly changing the files processed every time a new iteration starts is likely to cause confusion. It seems more natural that the default behavior would be to get a list of files and not change them as long as the object is used. This would match the behavior of LineSentence--if you change the contents of the file between iterations, you'll get different results, but you can't change the reference to the file after the object has been created.

I would personally be caught off guard if the files changed between iterations. While this could be useful in some cases, I think it is a risky default behavior. Adding some capabilities to do this, however, may make sense. But I'd rather not do that unless a compelling use case is presented.

What I've done instead is log the list of files read when the object is created at the info level, so there's some sort of explicit record available of what the object is reading.

gojomo commented on gensim/test/test_word2vec.py:
```diff
@@ -763,6 +763,15 @@ def testLineSentenceWorksWithNormalFile(self):
+        """Does LineSentencePath work with a path argument?"""
+        logging.debug(word2vec)
+        with utils.smart_open(datapath('lee_background.cor')) as orig:
+            sentences = word2vec.LineSentencePath(datapath('LineSentencePath'))
```
Perhaps rather than creating a custom new split (and duplication) of the `lee_background.cor`, a new subdir of test-data `lee` could be added, with just the `lee.cor` and `lee_background.cor` files. The test would make sure this new class on the directory yields the same set of docs as the *two* other files read individually. And eventually, the duplication of these files in two places could be eliminated by making all the other tests/demos/tutorials just grab the lee* files from this new canonical place. (That is: aim for less duplication in the long run by re-using the same files everywhere.)

michaelwsherman replied:
I thought about this, but I think making a lee directory may be worse than having a special PathLineSentences directory.

It seems there are quite a few lee* files in test_data already. I could move those two files you mentioned, but then we'd have a bunch of lee files still in the main test_data directory. Those files could be moved as well, except some of them aren't compatible with PathLineSentences (.bin, .vec) which meant that they couldn't be moved without breaking the test. Then we would have lee* files in two places, which could lead to confusion later on. There is also the issue that something that makes sense to add to this lee directory in the future could break the test.

Given that the PathLinesSentences class operates on a full directory (which makes it somewhat unique) I think it makes more sense to keep it's test dependent on a directory that isn't overloaded for use with other tests.

What I will do is use smaller files that aren't duplicated.

michaelwsherman replied:
Change made, put in a new tiny test corpus, and changed the test to read the files and combine them manually and compare to PathLineSentences, rather than having a single file duplicating the contents of the PathLineSentences folder in test_data

michaelwsherman commented:
I think I've addressed all the comments. Please re-review.

I'm having some problems with figuring out what travis-ci is angry about. I did find an inline comment missing a whitespace and fixed that. But I think the other comments are due to extra spaces in [i : i +...] line and that's elsewhere in the files. I also see errors with the new corpus file--not sure how to tell travis-ci to not treat those files as code. I'll review after the current style check is done and see if I can get to the bottom of things.

michaelwsherman commented:
Nevermind, looks like travis just checks my changes -- not code already in the repo. But I did fix another style issue elsewhere in the code. 

I think the style check should pass now.

michaelwsherman commented:
Cleaning up my noob git branching fail, no more dangling branches not actually getting used.

michaelwsherman commented:
@gojomo or other maintainers--Any other changes? Any questions? Please let me know. Thank you.

menshikh-iv commented:
Looks good for me, thank you @michaelwsherman, congratz with your first PR:1st_place_medal:

piskvorky commented on gensim/models/word2vec.py:
```diff
@@ -1517,7 +1517,57 @@ def __iter__(self):
+        self.limit = limit
+
+        if os.path.isfile(self.source):
+            logging.warning('single file read, better to use models.word2vec.LineSentence')
```
Should be `logger` (each module has its own logger in gensim).

Applies here and elsewhere in this PR.

piskvorky commented on gensim/models/word2vec.py:
```diff
@@ -1517,7 +1517,57 @@ def __iter__(self):
+            self.source = os.path.join(self.source, '')  # ensures os-specific slash at end of path
+            logging.debug('reading directory ' + self.source)
+            self.input_files = os.listdir(self.source)
+            self.input_files = [self.source + file for file in self.input_files]  # make full paths
```
`file` is a reserved keyword in Python. Better use `filename` or something like that.

piskvorky commented on gensim/models/word2vec.py:
```diff
@@ -1517,7 +1517,57 @@ def __iter__(self):
+        else:  # not a file or a directory, then we can't do anything with it
+            raise ValueError('input is neither a file nor a path')
+
+        logging.info('files read into PathLineSentences:' + '\n'.join(self.input_files))
```
Better to pass the formatting arguments as arguments: `logger.info("%s", y)` instead of `logger.info("%s" % y)` .

Here and elsewhere.

piskvorky commented on gensim/models/word2vec.py:
```diff
@@ -1517,7 +1517,57 @@ def __iter__(self):
+        """
+        `source` should be a path to a directory (as a string) where all files can be opened by the
+        LineSentence class. Each file will be read up to
+        `limit` lines (or no clipped if limit is None, the default).
```
`no` => `not`.

michaelwsherman commented:
Fixes from @piskvorky in PR #1573 .
